crypto-mcl.ts

```diff
@@ -0,0 +1,188 @@
+    mtr: IMasterTrace;
+}
+
+/**
+ * The request from the health authority to the location owner to create a proof to be sent to
+ * the visitors.
+ */
+interface IPreTrace{
+    Tprime: mcl.G1
+    info: string
+    nonce: Uint8Array
+    ctxt: Uint8Array
+}
+
+/**
+ * Section7 implements the cryptographic protocol in section 7 of the CrowdNotifier whitepaper.
+ */
+export class Section7{
+    /**
+     * Setup for the HealthAuthority - returns a keypair that can be used to encrypt the QRtrack codes.
+     */
+    static setupHA(): IKeyPair{
+        return crypto_box_keypair();
+    }
+
+    /**
+     * Creates the data necessary for a location owner.
+     * @param info is the public info of the location, e.g., name:location:room
+     * @param pkH the public key of the health authority
+     * @return the LocationData to be printed in 2 (3?) QRcodes
+     */
+    static genCode(info: string, pkH: Uint8Array): ILocationData{
+        const mskv = new mcl.Fr();
+        mskv.setByCSPRNG();
+        const mskHA = new mcl.Fr();
+        mskHA.setByCSPRNG();
+        const msk = mcl.add(mskv, mskHA);
+        const mpk = mcl.mul(mcl.baseG2(), msk);
+        const nonce = randombytes_buf(msk.serialize().byteLength * 2);
+        const mskHAEnc = crypto_box_seal(mskHA.serialize(), pkH);
+        const mtr = {mskv, info, nonce, mskHAEnc}
+        return {ent: mpk, pEnt: nonce, mtr};
+    }
+
+    /**
+     * Creates a private record for the user to store in her phone.
+     * @param entry as presented in the QRentry
+     * @param proofEntry as presented in the QRentry
+     * @param info as presented in the QRentry
+     * @param counter as hours since the unix epoch
+     * @param aux free string
+     * @return user record to be stored
+     */
+    static scan(entry: mcl.G2, proofEntry: Uint8Array, info: string, counter: number, aux: string): IUserRecord{
+        const r1 = new mcl.Fr();
+        r1.setByCSPRNG();
+        const r2 = new mcl.Fr();
+        r2.setByCSPRNG();
+        const k = crypto_secretbox_keygen();
+        const nonce = secretbox_nonce();
+        const c = crypto_secretbox_easy(from_string(aux), nonce, k);
+
+        const R1 = mcl.mul(mcl.baseG2(), r1);
+        const H1inc = this.infoNonceCounter(info, proofEntry, counter);
+        const Z1 = mcl.pairing(mcl.mul(H1inc, r1), entry);
+        const R2 = mcl.mul(mcl.baseG2(), r2);
+        const Z2 = mcl.pairing(mcl.mul(H1inc, r2), entry);
+        const K = mcl.xor(k, mcl.hashT(Z2));
+
+        return {R1, Z1, R2, K, c, nonce};
+    }
+
+    /**
+     * Convenience method to calculate the hash-to-G1 of the info, nonce, and counter.
+     * @param info
+     * @param nonce
+     * @param counter
+     */
+    static infoNonceCounter(info: string, nonce: Uint8Array, counter: number): mcl.G1 {
+        const counterStr = counter.toString();
+        const infoNonceCnt = new Uint8Array(info.length + nonce.length + counterStr.length);
```

Comment:
some as earlier discussion. This needs to be a bijectivec encoding of the three fields. So either take protobuf route again, or make sure to prepend lengths.

Reply:
I was hoping to get away with the nonce in the middle ;)

Added a protobuf